Crashlytics/Crashlytics/Controllers/FIRCLSMetricKitManager.m

```diff
@@ -0,0 +1,320 @@
+      [self writeThreadsToFile:&metricKitFile threads:threads];
+      [self writeMetadataToFile:&metricKitFile metadata:metadata];
+
+      NSDictionary *hangDict = @{@"hang_duration" : hangDiagnostic.hangDuration};
+      [self writeEventSpecificDataToFile:&metricKitFile event:@"hang" data:hangDict];
+
+      FIRCLSFileWriteSectionEnd(&metricKitFile);
+    }
+
+    if (hasCPUException) {
+      FIRCLSFileWriteSectionStart(&metricKitFile, "cpu_exception_event");
+      MXCPUExceptionDiagnostic *cpuExceptionDiagnostic =
+          [diagnosticPayload.cpuExceptionDiagnostics objectAtIndex:0];
+
+      NSData *threads = [cpuExceptionDiagnostic.callStackTree JSONRepresentation];
+      NSData *metadata = [cpuExceptionDiagnostic.metaData JSONRepresentation];
+      [self writeThreadsToFile:&metricKitFile threads:threads];
+      [self writeMetadataToFile:&metricKitFile metadata:metadata];
+
+      NSDictionary *cpuDict = @{
+        @"total_cpu_time" : cpuExceptionDiagnostic.totalCPUTime,
+        @"total_sampled_time" : cpuExceptionDiagnostic.totalSampledTime
+      };
+      [self writeEventSpecificDataToFile:&metricKitFile event:@"cpu_exception" data:cpuDict];
+
+      FIRCLSFileWriteSectionEnd(&metricKitFile);
+    }
+
+    if (hasDiskWriteException) {
+      FIRCLSFileWriteSectionStart(&metricKitFile, "disk_write_exception_event");
+      MXDiskWriteExceptionDiagnostic *diskWriteExceptionDiagnostic =
+          [diagnosticPayload.diskWriteExceptionDiagnostics objectAtIndex:0];
+
+      NSData *threads = [diskWriteExceptionDiagnostic.callStackTree JSONRepresentation];
+      NSData *metadata = [diskWriteExceptionDiagnostic.metaData JSONRepresentation];
+      [self writeThreadsToFile:&metricKitFile threads:threads];
+      [self writeMetadataToFile:&metricKitFile metadata:metadata];
+
+      NSDictionary *diskDict =
+          @{@"total_writes_caused" : diskWriteExceptionDiagnostic.totalWritesCaused};
+      [self writeEventSpecificDataToFile:&metricKitFile
+                                   event:@"disk_write_exception"
+                                    data:diskDict];
+
+      FIRCLSFileWriteSectionEnd(&metricKitFile);
+    }
+
+    // If this method was called because of a fatal event, fulfill the MetricKit promise so that
+    // uploading of the Crashlytics report continues. If not, the promise has already been
+    // fulfillled. Since we only want to handle one crash per report, if this was a fatal event
+    // return from the method.
+    if (fatal) {
+      @synchronized(self) {
+        [self.metricKitDataAvailable fulfill:nil];
+      }
+      return;
+    }
+  }
+}
+
+/*
+ * Required for MXMetricManager subscribers. Since we aren't currently collecting any MetricKit
+ * metrics, this method is left empty.
+ */
+- (void)didReceiveMetricPayloads:(NSArray<MXMetricPayload *> *)payloads {
+  FIRCLSDebugLog(@"[Crashlytics:Crash] received %d MetricKit metric payloads\n", payloads.count);
```

Comment:
Should we log if we're not doing anything with it? Will we or customers use this?

Reply:
Thinking we should remove the log cause it might confuse customers.